book-content/chapters/05-unions-literals-and-narrowing.md

```diff
@@ -0,0 +1,1856 @@
+
+```typescript
+const getAlbumYear = (year: string | number | boolean) => {
+  if (typeof year === "string") {
+    console.log(`The album was released in ${year}.`); // `year` is string
+  } else if (typeof year === "number") {
+    console.log(`The album was released in ${year}.`); // `year` is number | boolean
+  }
+
+  console.log(year); // `year` is string | number | boolean
+};
+```
+
+This is a powerful example of how TypeScript can read your runtime code and use it to narrow down the type of a value.
+
+### Other Ways to Narrow
+
+The `typeof` operator is just one way to narrow types.
+
+TypeScript can use other conditional operators like `&&` and `||`, and will take the truthiness into account for coercing the boolean value. It's also possible to use other operators like `instanceof` and `in` for checking object properties. You can even throw errors or use early returns to narrow types.
+
+We'll take a closer look at these in the following exercises.
+
+### Exercises
+
+#### Exercise 1: Narrowing with `if` Statements
+
+Here we have a function called `validateUsername` that takes in either a `string` or `null`, and will always return a `boolean`:
+
+```ts twoslash
+// @errors: 18047
+function validateUsername(username: string | null): boolean {
+  return username.length > 5;
+
+  return false;
+}
+```
+
+Tests for checking the length of the username pass as expected:
+
+```typescript
+it("should return true for valid usernames", () => {
+  expect(validateUsername("Matt1234")).toBe(true);
+
+  expect(validateUsername("Alice")).toBe(false);
+
+  expect(validateUsername("Bob")).toBe(false);
+});
+```
+
+However, we have an error underneath `username` inside of the function body, because it could possibly be `null` and we are trying to access a property off of it.
+
+```typescript
+it("Should return false for null", () => {
+  expect(validateUsername(null)).toBe(false);
+});
+```
+
+Your task is to rewrite the `validateUsername` function to add narrowing so that the `null` case is handled and the tests all pass.
+
+<Exercise title="Exercise 1: Narrowing with `if` Statements" filePath="/src/018-unions-and-narrowing/059-narrowing-with-if-statements.problem.ts"></Exercise>
+
+#### Exercise 2: Throwing Errors to Narrow
+
+Here we have a line of code that uses `document.getElementById` to fetch an HTML element, which can return either an `HTMLElement` or `null`:
+
+```typescript
+const appElement = document.getElementById("app");
+```
+
+Currently, a test to see if the `appElement` is an `HTMLElement` fails:
+
+```ts twoslash
+// @errors: 2344
+import { Equal, Expect } from "@total-typescript/helpers";
+
+const appElement = document.getElementById("app");
+
+// ---cut---
+type Test = Expect<Equal<typeof appElement, HTMLElement>>;
+```
+
+Your task is to use `throw` to narrow down the type of `appElement` before it's checked by the test.
```

Comment:
This has not been shown at all before so perhaps it might throw off the reader, I guess it really depends on the reader.